File: Ruby/Janken/janken.rb
def start_janken
	puts '「じゃんけん・・・」'
	puts '> 0.グー 1.チョキ 2.パー'
	input = gets.chomp.to_i
	# 入力値が期待値(0, 1, 2)通りか判定
	if (input != 0 &&
		input != 1 &&
		input != 2)
		# 期待通りでない場合、やり直し
		puts '入力値は「0, 1, 2」のいずれかを入力してください'
		start_janken
	else
		# 期待通りの場合、じゃんけんの勝敗を決める
		judgement_janken(input)
	end
end

def judgement_janken(player_idx)
	janken = ['グー', 'チョキ', 'パー']

	# プレイヤーの手
	player_hand = janken[player_idx]

	# コンピューターの手
	comp_idx = rand(0..2)
	comp_hand = janken[comp_idx]

	# 結果を表示
	puts '「ぽい！」'
	puts '＊コンピュータ：' + comp_hand
	puts '＊あなた ：' + player_hand
	if player_hand == comp_hand
		puts '「アイコでしょ！」'
		start_janken
	elsif (player_hand == janken[0] && comp_hand == janken[1] ||
				 player_hand == janken[1] && comp_hand == janken[2] ||
				 player_hand == janken[2] && comp_hand == janken[0])
		puts '「あなたの勝ち！」'
	else
		puts 'コンピューターの勝ち！'
	end
end

# メイン処理
start_janken

File: TypeScript/Janken/janken.ts
import * as readline from "node:readline";

function question(rl: readline.Interface, prompt: string): Promise<string> {
  return new Promise((resolve) => rl.question(prompt, resolve));
}

async function promptHandIndex(rl: readline.Interface): Promise<0 | 1 | 2> {
  const raw = (await question(rl, "> 0.グー 1.チョキ 2.パー : ")).trim();
  const n = Number.parseInt(raw, 10);
  if (n === 0 || n === 1 || n === 2) return n;
  console.log("入力値は 0, 1, 2 のいずれかを入力してください");
  return promptHandIndex(rl);
}

async function main(): Promise<void> {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  try {
    const hands = ["グー", "チョキ", "パー"] as const;

    while (true) {
      console.log("じゃんけん・・・");
      const playerIdx = await promptHandIndex(rl);
      const compIdx = Math.floor(Math.random() * 3) as 0 | 1 | 2;

      const playerHand = hands[playerIdx];
      const compHand = hands[compIdx];

      console.log("ぽん！！");
      console.log(`(コンピューター) ${compHand}`);
      console.log(`(あなた) ${playerHand}`);

      if (playerIdx === compIdx) {
        console.log("アイコでしょ！！");
        continue;
      }

      const playerWin =
        (playerIdx === 0 && compIdx === 1) || (playerIdx === 1 && compIdx === 2) || (playerIdx === 2 && compIdx === 0);

      console.log(playerWin ? "あなたの勝ち！！" : "コンピューターの勝ち！！");
      break;
    }
  } finally {
    rl.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exitCode = 1;
});
